libavutil/hwcontext_mediafoundation.c

```
@@ -0,0 +1,445 @@
+            goto fail;
+        }
+
+        priv->d3d9 = createD3D(D3D_SDK_VERSION);
+        if (!priv->d3d9) {
+            av_log(ctx, loglevel, "Failed to create IDirect3D object\n");
+            goto fail;
+        }
+
+        IDirect3D9_GetAdapterDisplayMode(priv->d3d9, hwctx->init_d3d9_adapter, &d3ddm);
+        d3dpp.Windowed         = TRUE;
+        d3dpp.BackBufferWidth  = 640;
+        d3dpp.BackBufferHeight = 480;
+        d3dpp.BackBufferCount  = 0;
+        d3dpp.BackBufferFormat = d3ddm.Format;
+        d3dpp.SwapEffect       = D3DSWAPEFFECT_DISCARD;
+        d3dpp.Flags            = D3DPRESENTFLAG_VIDEO;
+
+        hr = IDirect3D9_CreateDevice(priv->d3d9, hwctx->init_d3d9_adapter, D3DDEVTYPE_HAL, GetDesktopWindow(),
+                                    D3DCREATE_SOFTWARE_VERTEXPROCESSING | D3DCREATE_MULTITHREADED | D3DCREATE_FPU_PRESERVE,
+                                    &d3dpp, &d3d9device);
+        if (FAILED(hr)) {
+            av_log(ctx, loglevel, "Failed to create Direct3D device\n");
+            goto fail;
+        }
+    }
+
+    priv->dxva2_dll = LoadLibraryW(L"dxva2.dll");
+    if (!priv->dxva2_dll) {
+        av_log(ctx, loglevel, "Failed to load DXVA2 library\n");
+        goto fail;
+    }
+        createDeviceManager = (pCreateDeviceManager9 *)GetProcAddress(priv->dxva2_dll, "DXVA2CreateDirect3DDeviceManager9");
+    if (!createDeviceManager) {
+        av_log(ctx, loglevel, "Failed to locate DXVA2CreateDirect3DDeviceManager9\n");
+        goto fail;
+    }
+
+    hr = createDeviceManager(&resetToken, &hwctx->d3d9_manager);
+    if (FAILED(hr)) {
+        av_log(ctx, loglevel, "Failed to create Direct3D device manager\n");
+        goto fail;
+    }
+
+    hr = IDirect3DDeviceManager9_ResetDevice(hwctx->d3d9_manager, d3d9device, resetToken);
+    if (FAILED(hr)) {
+        av_log(ctx, loglevel, "Failed to bind Direct3D device to device manager\n");
+        goto fail;
+    }
+
```

Does the pointer d3d9device need to be tested like the one in goto fail?

---

When IDirect3D9_CreateDevice returns, the HRESULT is checked.  If it passed, we can rely on d3d9device being valid.  I believe this should be ok, unless I am missing something else here?

---

Yes it seems ok, IDirect3DDeviceManager9_ResetDevice should fail if d3d9device is invalid anyway.
I was more "worried" about `d3d9device = hwctx->init_d3d9_device;`